cpp/inc/bond/core/maybe.h

```diff
@@ -14,141 +19,556 @@ namespace bond
+    {
+        return lhs._value == rhs;
+    }
+
+    /// @brief Compares a maybe and a value for inequality.
+    ///
+    /// See operator==(const maybe_common&,const T&) for details about how
+    /// maybes holding nothing are handled.
+    ///
+    /// @since 8.0.0
+    friend bool operator!=(const maybe_common& lhs, const T& rhs)
+    {
+        return !(lhs == rhs);
+    }
+
+    /// @brief Compares a value and a maybe for equality.
+    ///
+    /// @return false if the maybe holds nothing; otherwise, calls
+    /// operator== with the provided value and maybe's value.
+    ///
+    /// @since 8.0.0
+    friend bool operator==(const T& lhs, const maybe_common& rhs)
+    {
+        return lhs == rhs._value;
+    }
+
+    /// @brief Compares and a value and a maybe for inequality.
+    ///
+    /// See operator==(const T&,const maybe_common&) for details about how
+    /// maybes holding nothing are handled.
+    ///
+    /// @since 8.0.0
+    friend bool operator!=(const T& lhs, const maybe_common& rhs)
+    {
+        return !(lhs == rhs);
+    }
+
+protected:
+    maybe_common() = default;
+
+    #if defined(_MSC_VER) && _MSC_VER < 1900
+    // Using = default with MSVC 2013 on this function causes the compiler
+    // to make extra copies, which makes this type fail when it's holding a
+    // move-only type.
+    //
+    // Manually implementing this function works just fine, however.
+    maybe_common(const maybe_common& that)
+        : _value(that._value)
+    { }
+    #else
+    maybe_common(const maybe_common&) = default;
+    #endif
+
+    template <typename... Args>
+    explicit maybe_common(const T& value, Args&&... args)
+        : _value()
+    {
+        _value.emplace(value, std::forward<Args>(args)...);
+    }
+
+    template <typename... Args>
+    explicit maybe_common(T&& value, Args&&... args)
+        : _value()
+    {
+        _value.emplace(std::move(value), std::forward<Args>(args)...);
+    }
+
+    maybe_common(maybe_common&& that) BOND_NOEXCEPT_IF(std::is_nothrow_move_constructible<boost::optional<T>>::value)
+        : _value(std::move(that._value))
+    {
+        // unlike std::optional/boost::optional, moved-from bond::maybe
+        // instances are guaranteed to be nothing.
+        //
+        // asigning boost::none is noexcept, but assigning { } is not
+        that._value = boost::none;
+    }
+
+    #if defined(_MSC_VER) && _MSC_VER < 1900
+    // Using = default with MSVC 2013 on this function causes the compiler
+    // to make extra copies, which makes this type fail when it's holding a
+    // move-only type.
+    //
+    // Manually implementing this function works just fine, however.
+    maybe_common& operator=(const maybe_common& that)
+    {
+        _value = that._value;
+        return *this;
+    }
+
+    // MSVC 2013 cannot = default rvalue asignment operators
+    maybe_common& operator=(maybe_common&& that) BOND_NOEXCEPT_IF(BOND_NOEXCEPT_EXPR(_value = std::move(that._value)))
```

Comment:
The `BOND_NOEXCEPT_EXPR` doesn't help much here, just use `std::is_nothrow_move_assignable`?

Reply:
✔️ 
I tried to add tests for these, but the MSVC 2013 implementation is missing a bunch of the type traits. I could have used to Boost wrappers, but then the 2013 STL is missing `noexcept` specifiers on a lot of its types. So, for 2013, full `noexcept` support doesn't seem worth it...